### PR TITLE
Fix gradient accumulation

### LINKED ISSUE #1062: AdaScale with gradient accumulation: Sum or average of gradients?
Hi!

I am wondering what is the intended way to implement gradient accumulation when using AdaScale. Usually, for gradient accumulation, I would use a training loop like this

    for i, (X, y) in enumerate(train_loader):
        if (i+1) % num_accum_steps == 0:
            loss = loss_fn(model(X), y) / num_accum_steps
            loss.backward()
            opt.step()
            opt.zero_grad()
        else:
            with model.no_sync():
                loss = loss_fn(model(X), y) / num_accum_steps
                loss.backward()

That is, when the optimizer step is applied, the gradient stored in `p.grad` is the **average** gradient across accumulation steps. However, this [line](https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/adascale.py#L460) in the AdaScale code makes me think that AdaScale expects `p.grad` to contain the **sum** of gradients across accumulation steps.

Could somebody clarify what's the intended way?

Thanks!

### PR DESCRIPTION
**Pull request by cyugao**
Add ``is_scaled_loss`` flag to support both scaled / unscaled loss Fix ``test_grad_accum`` and``test_set_num_gradients_to_accumulate``

## What does this PR do?
Fixes #1062.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/optim/adascale.py [hunk 1]
```diff
@@ -136,6 +136,10 @@ class AdaScale(Optimizer):
             for smoothing and mu and sigma variables. False will
             use the method in the paper's Appendix B.3.
             Default: True, which is what have been validated so far.
+        is_scaled_loss (bool):
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:139]: Can you add to line 132 above and refer to this option so that people know there are 2 ways to do grad_accum?

File: fairscale/optim/adascale.py [hunk 2]
```diff
@@ -453,17 +459,19 @@ def _final_callback(self) -> None:
...
-            # np array doesn't support /=.
-            total_grad_sqr = total_grad_sqr / (self._num_grads_to_accum**2)
 
         # Wait for all_reduce to be done and move it to cpu & np.
         if work:
             work.wait()
         local_grad_sqr = self._local_grad_sqr.cpu().numpy()
 
+        if self._num_grads_to_accum > 1:
+            # Handle scaling for for gradient accumulation
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:469]: Can you expand this comment and document the reason behind line 471 and 473? It is not easy to understand and we may forget its reasons without detailed explanation here. :)

File: fairscale/optim/adascale.py [hunk 3]
```diff
@@ -453,17 +459,19 @@ def _final_callback(self) -> None:
...
+            if self._is_scaled_loss:
+                local_grad_sqr *= self._num_grads_to_accum**2
+            else:
+                total_grad_sqr /= self._num_grads_to_accum**2
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:473]: interesting, `/=` not worked in the past, maybe newer numpy works with it now?

File: fairscale/optim/adascale.py [hunk 4]
```diff
@@ -509,7 +517,12 @@ def step(self, *args: Any, **kwargs: Any) -> Optional[float]:
         original_lr = []
         for idx, param_group in enumerate(self._optimizer.param_groups):
             original_lr.append(param_group["lr"])
-            param_group["lr"] = self.gain(pg_idx=idx) * param_group["lr"]
+            if self._num_grads_to_accum > 1 and not self._is_scaled_loss:
+                # Divide by num_grads_to_accum to account for gradient accumulation.
```
Comment (min-xu-ai) [inline fairscale/optim/adascale.py:521]: this comment only states the same info as the code, which is not very useful. Let's explain that when scaled loss is not used, the gain computed is correct, but in order to match what DDP does, we need to scale the grad back by N.

However, with momentum, scale LR by N here is not the same as scale the param.grad by N in the DDP case, right?

File: tests/optim/test_single_node_adascale.py [hunk 5]
```diff
@@ -194,16 +195,23 @@ def test_lr_scheduler():
 
 @skip_if_no_cuda
 @pytest.mark.parametrize("debias_ewma", [True, False])
-def test_add_param_group(debias_ewma):
-    """Test AdaScale supports add_param_group() API."""
+@pytest.mark.parametrize("is_scaled_loss", [True, False])
```
Comment (min-xu-ai) [inline tests/optim/test_single_node_adascale.py:198]: very nice to introduce a bool for this

File: tests/optim/test_single_node_adascale.py [hunk 6]
```diff
@@ -82,14 +82,14 @@ def test_grad_accum(test_case, cpu):
         in_data_0 = Tensor(in_data[0])
         if not cpu:
             in_data_0 = in_data_0.cuda()
-        out = model(in_data_0)
-        out.sum().backward()
+        loss = model(in_data_0).sum() / 2
```
Comment (min-xu-ai) [inline tests/optim/test_single_node_adascale.py:85]: in this test, can we add expected grad to adascale_test_data on line 63? That would explicitly check that the grad in all three cases below are the same:
1. ddp
2. grad with loss scale
3. grad without loss scale but with a final grad avg step

This should ensure there is no bug I suspected above in the step() function.

**Review by min-xu-ai — COMMENTED** [review]
Very nice! I have a few minor suggestions and a one question below.

File: fairscale/optim/adascale.py (hunk 3 above)
Comment (cyugao) [reply fairscale/optim/adascale.py:473]: I think if you use `a /= 2` when `a` is an integer array it will break since the result will be float, otherwise it should be fine?

**Review by min-xu-ai — APPROVED** [review]
LGTM! Good luck merge it.
